sql: panic: invalid datum type given: int, expected unknown

```
CREATE TABLE x (a INT8);

CREATE TABLE y (b INT8);

INSERT INTO x VALUES (1);

INSERT INTO y VALUES (2);

SELECT a, (SELECT a FROM y) FROM x;
```

```
panic: invalid datum type given: int, expected unknown [recovered]
	panic: panic while executing 1 statements: SELECT _, (SELECT _ FROM _) FROM _; caused by invalid datum type given: int, expected unknown

goroutine 368 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0015d6000, 0x3a6b700, 0xc00163e840, 0x2deac40, 0xc0013dd950)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:721 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0015d6000, 0x3a6b700, 0xc00163e840)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:434 +0x61
panic(0x2deac40, 0xc0013dd950)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xc0013c3a68, 0x3aa2980, 0xc001344a80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:172 +0x1e6
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc0013ec500, 0xc0015d6350, 0x3a6c8c0, 0xc0013f0400, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:183 +0x4fd
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x3a6b7c0, 0xc0013d1d40, 0x3a75b00, 0xc0013ec500, 0x3a5ce40, 0xc0013f8380)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:175 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc0013ec500, 0x3a6b7c0, 0xc0013d1d40)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:808 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc0013f6000, 0x3a6b7c0, 0xc0013d1d40, 0x3452fc0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:626 +0x1e9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000444a00, 0xc0013e2b40, 0xc0013cce10, 0xc0018f08c8, 0xc000bc8dc0, 0xc0015d6410, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:253 +0x8a0
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000444a00, 0x3a6b7c0, 0xc0013d15f0, 0xc0015d6410, 0xc0013e2b40, 0xc0013cce10, 0x3a6d4c0, 0xc0013c37a0, 0xc000bc8dc0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:841 +0x227
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc0015d6000, 0x3a6b7c0, 0xc0013d15f0, 0xc0015d6350, 0x3, 0x7f539186d2a8, 0xc000a7bce0, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:895 +0x283
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0015d6000, 0x3a6b7c0, 0xc0013d15f0, 0xc0015d6350, 0x7f539186d2a8, 0xc000a7bce0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:731 +0x658
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0015d6000, 0x3a6b7c0, 0xc0013d15f0, 0x3a70fc0, 0xc0013b1c80, 0xc0016dc2f5, 0x22, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:412 +0xaf1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0015d6000, 0x3a6b7c0, 0xc0013d15f0, 0x3a70fc0, 0xc0013b1c80, 0xc0016dc2f5, 0x22, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:100 +0x610
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0015d6000, 0x3a6b7c0, 0xc0013d15c0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1200 +0x1ad8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0015d6000, 0x3a6b700, 0xc00163e840, 0xc0001110b8, 0x5400, 0x15000, 0xc000111150, 0xc00032c040, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1136 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc00095d340, 0x3a6b700, 0xc00163e840, 0xc0015d6000, 0x5400, 0x15000, 0xc000111150, 0xc00032c040, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:436 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0002dcff9, 0xc0002c8bc0, 0x3a6b700, 0xc00163e840, 0xc00032c040, 0xc00095d340, 0xc00175a000, 0x3a6efc0, 0xc0002c8b80, 0xc00165c0c0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21f
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:515 +0x17b
```

--- a/pkg/sql/logictest/testdata/logic_test/apply_join
+++ b/pkg/sql/logictest/testdata/logic_test/apply_join
@@ -268,3 +268,13 @@ LIMIT
     89:::INT8;
 ----
 true
+
+# Regression test for #37454: untyped null produced at top level.
+
+statement ok
+CREATE TABLE x (a INT8); CREATE TABLE y (b INT8); INSERT INTO x VALUES (1); INSERT INTO y VALUES (2);
+
+query II
+SELECT a, (SELECT a FROM y) FROM x
+----
+1  1

--- a/pkg/sql/opt/exec/execbuilder/scalar_builder.go
+++ b/pkg/sql/opt/exec/execbuilder/scalar_builder.go
@@ -143,7 +143,11 @@ func (b *Builder) indexedVar(
 	idx, ok := ctx.ivarMap.Get(int(colID))
 	if !ok {
 		if b.nullifyMissingVarExprs > 0 {
-			return tree.DNull
+			expr, err := tree.ReType(tree.DNull, md.ColumnMeta(colID).Type)
+			if err != nil {
+				panic(pgerror.AssertionFailedf("unexpected failure during ReType: %v", err))
+			}
+			return expr
 		}
 		panic(pgerror.AssertionFailedf("cannot map variable %d to an indexed var", log.Safe(colID)))
 	}

